### PR TITLE
Adjust utilization sheet for printing

### PR DESCRIPTION
**Pull request by pzehner**
This PR aims to have a printable utilization cheat sheet. It may look messy, as different things are made here:

- Compress the *content* of `utilization.md`: most of the extra, introductory, or too verbose text blocs were removed or shortened. As the cheat sheet is intended for people with a minimum Kokkos knowledge, cutting the introductions down seems acceptable. Some sections were re-visited, like the Initialization, the Kokkos concepts, the Mirror views, the Fences, and the Hierarchical parallelism.
- Manage *how* to handle `utilization.md`: this is more about updating the existing scripts that were developed for `install.md` only to manage both files now. The Pandoc filter was updated to better detect lines of text from lines of code.

The resulting PDF file is now 5 pages long.

### CODE REVIEW

File: convert.sh [hunk 1]
```diff
@@ -48,20 +48,24 @@ convert () {
 }
 
 patch_modifs () {
-    local file="$1"
+    local output_file_diff="$1"
```
Comment (cedricchevalier19) [inline convert.sh:51]: Great idea!

File: utilization.md [hunk 2]
```diff
@@ -1,142 +1,196 @@
...
-### Initialize and finalize Kokkos
+### Scope guard
 
 ```cpp
-Kokkos::initialize(int& argc, char* argv[]);
-{
+int main(int argc, char* argv[]) {
+    Kokkos::ScopeGuard kokkos();
     /* ... */
+    return 0;
```
Comment (cedricchevalier19) [inline utilization.md:109]: Still not sure if it worth mentioning it.

File: utilization.md [hunk 3]
```diff
@@ -161,48 +215,34 @@ Kokkos::View<double**[25]> view3D("view3D", 50, 42, 25);
...
 
-| Layout                   | Description                                                     | Other name                   | Efficient for |
-|--------------------------|-----------------------------------------------------------------|------------------------------|---------------|
-| `Kokkos::LayoutRight()`  | Strides increase from the right most to the left most dimension | row-major or C-like          | CPU           |
-| `Kokkos::LayoutLeft()`   | Strides increase from the left most to the right most dimension | column-major or Fortran-like | GPU           |
-| `Kokkos::LayoutStride()` | Strides can arbitrary for each dimension                        |                              |               |
-
-If no layouts are specified, the most efficient one for the memory space is used.
-For performance, memory access patterns must result in caching on a CPU and coalescing on a GPU.
+By default, the most efficient layout for the current memory space is used.
```
Comment (cedricchevalier19) [inline utilization.md:245]: A little bit strong. The default is a reasonable choice for loops on the high frequency index.

File: utilization.md [hunk 4]
```diff
@@ -379,55 +349,36 @@ A `HostMirror` view is allocated in the host memory space as a mirror of a devic
...
 
-A `subview` is a `View` that is a subset of another view that mimics the behavior of languages like Python or Fortran.
-
-<img title="Warning" alt="Warning" src="./images/warning_txt.svg" height="25"> A subview has the same reference count as its parent View, so the parent View won’t be deallocated before all subviews go away.
-
-<img title="Warning" alt="Warning" src="./images/warning_txt.svg" height="25"> Every subview is also a view. This means that you may take a subview of a subview.
-
 <!--#ifndef PRINT-->
+<img title="Warning" alt="Warning" src="./images/warning_txt.svg" height="25"> A subview has the same reference count as its parent view, so the parent view won’t be deallocated before all subviews go away.
+
```
Comment (cedricchevalier19) [inline utilization.md:374]: Why is it a warning? It looks like a nice feature to have.

File: utilization.md [hunk 5]
```diff
@@ -379,55 +349,36 @@ A `HostMirror` view is allocated in the host memory space as a mirror of a devic
 - [Kokkos Tutorials - Exercise 3](https://github.com/kokkos/kokkos-tutorials/blob/main/Exercises/03/Solution/exercise_3_solution.cpp)
 <!--#endif-->
 
-#### Create and allocate
+#### Always create and allocate
```
Comment (cedricchevalier19) [inline utilization.md:352]: Not clear what `create` means and how `always create` is different.

File: utilization.md [hunk 6]
```diff
@@ -715,80 +683,54 @@ Kokkos::parallel_for(
...
 ### Scratch memory space
 
-| Space level | Memory size                 | Access speed |
-|-------------|-----------------------------|--------------|
-| 0           | Limited (tens of kilobytes) | Fast         |
-| 1           | Larger (few gigabytes)      | Medium       |
+| Level | Memory size                 | Access speed |
+|-------|-----------------------------|--------------|
+| 0     | Limited (tens of kilobytes) | Fast         |
+| 1     | Larger (few gigabytes)      | Slow         |
```
Comment (cedricchevalier19) [inline utilization.md:733]: Still better than non-scratch, `medium` was kind of more informative.

**Review by cedricchevalier19 — COMMENTED** [review]
I am a little bit annoyed by the removal of the small description of each section.

File: convert.sh (hunk 1 above)
Comment (pzehner) [reply convert.sh:51]: Self documented variables ftw!

File: utilization.md (hunk 2 above)
Comment (pzehner) [reply utilization.md:111]: Ok, let's comment it for now before this feature is tested.

File: utilization.md (hunk 5 above)
Comment (pzehner) [reply utilization.md:372]: Maybe "Create and always allocate" instead? This is to make a difference with the different versions, where the allocation is conditional.

Comment (cedricchevalier19) [reply utilization.md:372]: Looks clearer.

**Comment by pzehner** [comment]
> I am a little bit annoyed by the removal of the small description of each section.

This is understandable. But I think people reading the cheat sheet know what they are looking for.

File: utilization.md [hunk 7]
```diff
@@ -1,142 +1,198 @@
...
-| `-kokkos-threads=INT`    | Specify total number of threads or number of threads per NUMA region if used in conjunction with `--kokkos-numa` option                             |
-| `-kokkos-numa=INT`       | Specify number of NUMA regions used by process                                                                                                      |
-| `-device-id=INT`         | Specify device ID to be used by Kokkos                                                                                                              |
-| `-num-devices=INT[,INT]` | Specify the number of devices per node to be used when running MPI jobs; the optional second argument allows for an existing device to be ignored |
+| Execution space                     | Device backend | Host backend |
+|-------------------------------------|----------------|--------------|
+| `Kokkos::DefaultExecutionSpace`     | On device      | On host      |
+| `Kokkos::DefaultHostExecutionSpace` | On host        | On host      |
+| `Kokkos::Serial`                    | On host        | On host      |
```
Comment (cedricchevalier19) [inline utilization.md:123]: Why this backend?

File: utilization.md [hunk 8]
```diff
@@ -1,142 +1,198 @@
...
 
-#### Struct arguments
+`Kokkos::Serial` is used for debug purpose only.
```
Comment (cedricchevalier19) [inline utilization.md:125]: Is it always available?

File: utilization.md [hunk 9]
```diff
@@ -1,142 +1,198 @@
...
+| `Kokkos::DefaultExecutionSpace::memory_space`     | On device      | On host      |
+| `Kokkos::DefaultHostExecutionSpace::memory_space` | On host        | On host      |
+
+#### Specific memory spaces
+
+| Memory space                 | Description                                                                  |
+|------------------------------|------------------------------------------------------------------------------|
+| `Kokkos::HostSpace`          | Accessible by the host but not directly by the device                        |
+| `Kokkos::SharedSpace`        | Accessible by the host and the device; synchronization managed by the driver |
+| `Kokkos::ScratchMemorySpace` | Accessible by the team that created it and nothing else                      |
```
Comment (cedricchevalier19) [inline utilization.md:150]: No thread level scratch? https://kokkos.org/kokkos-core-wiki/API/core/policies/TeamHandleConcept.html?highlight=scratch#_CPPv4NK17TeamHandleConcept14thread_scratchEi

File: utilization.md [hunk 10]
```diff
@@ -161,141 +217,77 @@ Kokkos::View<double**[25]> view3D("view3D", 50, 42, 25);
...
-```cpp
-for (int i = 0; i < 10; i++) {
-    for (int j = 0; j < 50; ++j) {
-        view(i, j) = i + j;
-    }
-}
-```
+| Method        | Description                                                                                                           |
+|---------------|-----------------------------------------------------------------------------------------------------------------------|
+| `(i, j...)`   | Returns and sets the value                                                                                            |
```
Comment (cedricchevalier19) [inline utilization.md:224]: Access the value indexed by (i...)

File: utilization.md [hunk 11]
```diff
@@ -161,141 +217,77 @@ Kokkos::View<double**[25]> view3D("view3D", 50, 42, 25);
...
+| `size()`      | Returns the total number of elements in the view                                                                      |
+| `rank()`      | Returns the number of dimensions                                                                                      |
+| `layout()`    | Returns the layout of the view                                                                                        |
+| `extent(dim)` | Returns the number of elements in the desired dimension                                                               |
```
Comment (cedricchevalier19) [inline utilization.md:228]:
```suggestion
| `extent(dim)` | Returns the number of elements in the requested dimension                                                               |
```

**Review by cedricchevalier19 — APPROVED** [review]
I think we should review the content on paper later, it would be simpler to read than the diff!
Thanks for the opportunity to easily print it.

File: utilization.md (hunk 9 above)
Comment (pzehner) [reply utilization.md:150]: I'll modify it to include the thread as well.

File: utilization.md (hunk 10 above)
Comment (pzehner) [reply utilization.md:224]: To be honest, I'm not happy by calling `(i, j...)` a method. I can follow your comment to make it consistent though.

File: utilization.md (hunk 7 above)
Comment (pzehner) [reply utilization.md:123]: This was part of the original version. I'm not sure either why it's here, unless for explicit debug purpose.

File: utilization.md (hunk 8 above)
Comment (pzehner) [reply utilization.md:125]: The documentation says it's enabled by default unless a different host backend is used. So, I guess it is, but I may be wrong.

Comment (pzehner) [reply utilization.md:125]: Oh, right, I made a confusion between the backend and the execution space. If you select the OpenMP backend, is the Serial execution space still available?

Comment (pzehner) [reply utilization.md:125]: The answer is no.

So, either way, we specify that the Serial backend must be used, or we remove it from the section completely.

File: utilization.md (hunk 7 above)
Comment (pzehner) [reply utilization.md]: After discussion with Mathieu, we decided to remove this execution space.

**Pull request merged**